### PR TITLE
Add API to Request URL to OpenAPI Docs for a Request Header

### PR DESCRIPTION
**Pull request by Snehil-Shah**
Add a Method that returns URL to OpenAPI Docs for a Request Header String

Prerequisite for https://github.com/qdrant/qdrant-web-ui/pull/142

### CODE REVIEW

File: src/request-docs.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        if(matchReg.test(endpoint)){
+            if (method.toLowerCase() in openapi.paths[path]){
+                const openapiDocs = openapi.paths[path][method.toLowerCase()];
+                const docsURL = 'https://qdrant.github.io/qdrant/redoc/index.html#tag/' + openapiDocs.tags[0] + '/operation/' + openapiDocs.operationId;
+                return docsURL;
+            }
+        }
+    }
+    return null
+}
```
Comment (kartik-gupta-ij) [inline src/request-docs.js]: Try to use `OpenAPIExtractor` for parsing.

File: src/request-docs.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+const Methods = ['POST', 'GET', 'PUT', 'DELETE', 'PATCH', 'HEAD'];
+
+function matchRequest(openapi, method, endpoint){
+    const paths = Object.keys(openapi.paths);
+    for (const path of paths){
+        const matchReg = new RegExp('^/?' + path.slice(1,).replace(/{.*?}/g, '[-a-zA-Z0-9_<>]+') + '$');
+        if(matchReg.test(endpoint)){
+            if (method.toLowerCase() in openapi.paths[path]){
+                const openapiDocs = openapi.paths[path][method.toLowerCase()];
+                const docsURL = 'https://qdrant.github.io/qdrant/redoc/index.html#tag/' + openapiDocs.tags[0] + '/operation/' + openapiDocs.operationId;
```
Comment (kartik-gupta-ij) [inline src/request-docs.js]: set `docsURL` as an external parameter.

**Comment by kartik-gupta-ij** [comment]
Write test for Docs url fetching.

File: src/request-docs.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                return docsURL;
+            }
+        }
+    }
+    return null
+}
+
+export function getRequestDocs(openapi, requestString){
```
Comment (kartik-gupta-ij) [inline src/request-docs.js]: Convert this into a JavaScript class.

**Review by kartik-gupta-ij — CHANGES_REQUESTED** [review]

**Comment by Snehil-Shah** [comment]
@kartik-gupta-ij Updated the PR!

File: src/request-docs.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    constructor(docsBaseURL, openapi){
+        this.DOCS_BASE_URL = docsBaseURL;
+        this.openapi = openapi;
+        this.extractor = new OpenAPIExtractor(openapi);
+        this.methods = this.extractor.getMethodDefinitions();
+    }
+
+    matchRequest(method, endpoint){
+        for (const request of this.methods){
+            const matchReg = new RegExp('^/?' + request.path.slice(1,).replace(/{.*?}/g, '[-a-zA-Z0-9_<>]+') + '$');
```
Comment (kartik-gupta-ij) [inline src/request-docs.js]: Try using something similar `AutocompleteTrie` in place of `RegExp`.

File: src/request-docs.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        let tokens = tokenizeHeader(requestString);
+        if(tokens.length < 2 || requestString.slice(0,2) == '//'){
+            return null;
+        }
+        const method = tokens[0];
+        const endpoint = tokens.slice(1).join('/')
+        if (!this.HTTPMethods.includes(method)){
+            return null
+        }
+        return this.matchRequest(method.toLowerCase(), endpoint);
```
Comment (kartik-gupta-ij) [inline src/request-docs.js]: Make this similar to [completeRequestHeader](https://github.com/qdrant/autocomplete-openapi/blob/0d7287fceb939bf986b9d0bfe756c711de35c95c/src/autocomplete.js#L36) .
That will take tokens as argument and return `DOC_URL` by using similar method to `trieCompletion.addPath` and `trieCompletion.autocomplete`.

**Review by kartik-gupta-ij — CHANGES_REQUESTED** [review]

File: src/request-docs.js (hunk 5 above)
Comment (Snehil-Shah) [reply src/request-docs.js]: I did some refactoring, is the current implementation okay?

File: src/request-docs.js (hunk 4 above)
Comment (Snehil-Shah) [reply src/request-docs.js]: matchReg is just a regular expression to match paths, how can I use a Trie here?

**Comment by kartik-gupta-ij** [comment]
Use this
[refactor.patch](https://github.com/qdrant/autocomplete-openapi/files/14008004/refactor.patch)

**Comment by Snehil-Shah** [comment]
> Use this [refactor.patch](https://github.com/qdrant/autocomplete-openapi/files/14008004/refactor.patch)

Done! Thank You!

**Review by kartik-gupta-ij — APPROVED** [review]

File: src/request-docs.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            });
+        }
+    }
+
+    getRequestDocs(requestString){
+        let tokens = tokenizeHeader(requestString);
+        let terminal = this.trieCompletion.match(tokens);
+        if(terminal){
+            const docsURL = this.DOCS_BASE_URL + terminal.tags[0] + '/operation/' + terminal.operationId;
+            return docsURL;
```
Comment (kartik-gupta-ij) [inline src/request-docs.js]: Hey @Snehil-Shah
Can you removing this and just return `terminal`.
For example:
```
apiDocs.getRequestDocs('PATCH collections/test')
```
Which returns 'host/collections/operation/update_collection'
This url is too exact, we can't guarantee that in our documentation there will be exactly this collections ->operation -> update_collection path.

Comment (Snehil-Shah) [reply src/request-docs.js]: @kartik-gupta-ij is it okay now?

Comment (kartik-gupta-ij) [reply src/request-docs.js]: seems good to me!

**Review by generall — APPROVED** [review]

**Pull request merged**